addon/components/bs4/bs-toast.js

```diff
@@ -0,0 +1,315 @@
+   * @type {?Element}
+   * @private
+   */
+  toastElement = null;
+
+  /**
+   * The current state of the toast. Possible states are:
+   * - showing
+   * - shown
+   * - hiding
+   * - hidden
+   *
+   * @property state
+   * @type {String}
+   * @private
+   */
+  get state() {
+    let { toastElement } = this;
+    assert('Toast element must be set before state of toast could be determined', toastElement);
+
+    if (toastElement.classList.contains(ClassName.SHOWING)) {
+      return 'showing';
+    }
+
+    if (toastElement.classList.contains(ClassName.SHOW)) {
+      return 'shown';
+    }
+
+    if (toastElement.classList.contains(ClassName.HIDE)) {
+      return 'hidden';
+    }
+
+    // There is no hiding class. Instead hiding state is determined by not having any of these
+    // state classes. `ClassName.SHOW` is removed on the beginning of hiding animation and
+    // `ClassName.HIDE` not added before the end of the hiding animation.
+    return 'hiding';
+  }
+
+  /**
+   * Shows the toast.
+   *
+   * Should be executed as soon as the toast is inserted into DOM.
+   *
+   * @method _show
+   * @param {?Element} toastElement
+   * @return void
+   * @private
+   */
+  @action
+  _show() {
+    let { toastElement } = this;
+
+    // toast must not be shown if `show` argument is false
+    if (!this.show) {
+      return;
+    }
+
+    // toast is only showable if it's currently hidden or hiding
+    if (!['hiding', 'hidden'].includes(this.state)) {
+      return;
+    }
+
+    toastElement.classList.remove(ClassName.HIDE);
+
+    if (this.animate) {
+      toastElement.classList.add(ClassName.SHOWING);
+
+      let transitionDuration = getTransitionDurationFromElement(toastElement);
+      let timer = later(this, '_onShown', transitionDuration);
+      this._timers.set('onShown', timer);
+    } else {
+      this._onShown();
+    }
+  }
+
+  /**
+   * Executed at the end of show animation.
+   *
+   * @method _onShown
+   * @return void
+   * @private
+   */
+  _onShown() {
+    let { toastElement } = this;
+
+    toastElement.classList.remove(ClassName.SHOWING);
+    toastElement.classList.add(ClassName.SHOW);
+
+    if (typeof this.onShown === 'function') {
+      this.onShown();
+    }
+
+    this._timers.delete('onShown');
+
+    if (this.autohide) {
+      let timer = later(this, '_hide', toastElement, this.delay);
+      this._timers.set('hide', timer);
+    }
+  }
+
+  /**
+   * Hide the toast.
+   *
+   * @method _hide
+   * @return void
+   * @private
+   */
+  @action
+  _hide() {
+    let { toastElement } = this;
+
+    // a toast could only be hidden if in show or showing state
+    if (!['shown', 'showing'].includes(this.state)) {
+      return;
+    }
+
+    // if toast is in showing state the animation must be canceled
+    if (this.state === 'showing') {
+      cancel(this._timers.onShown);
+      this._timers.delete('onShown');
+    }
+
+    if (typeof this.onHide === 'function') {
+      this.onHide();
```

Comment:
For Modals we allow returning `false` to prevent hiding it, in case the user wants to apply some custom logic. Does it make sense to allow that here also?